### DISCUSSION TITLE
Enums, are they necessary?

### DISCUSSION THREAD

**Discussion by dixpac** [josefarias/hotwire_combobox#148] [Ideas]
HotwiredCombobox has a support for enums
`<%= combobox_tag "rating", Movie.ratings %>`

At the first glance api looks really nice and concise, but I think it has its drawbacks

1. Enums are part of the [ActiveRecord](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/enum.rb), which means if we have a form that is not directly backed by ActiveRecord, for example a FormObject, gem will fail to render `form.commbobox ....` 
<img width="1428" alt="Screenshot 2024-04-17 at 18 23 07" src="https://github.com/josefarias/hotwire_combobox/assets/7427365/3281ce76-cd98-4a24-a032-d2261ce258cd">

Error image above is a simple dummy form object from the repo that looks like this:

```ruby
# Form object dummy example

class State::Form
  include ActiveModel::Model

  attr_accessor :state_id

  def initialize(state_id: nil)
    @state_id = State.first
  end

  def save
    true
  end
end

```

```erb
<%= form_with model: @state_form, url: "#" do |form| %>
  <%= form.combobox "state", @state_options, id: "state-field" %>
  <%= form.submit %>
<% end %>

```

2. Localization, if the app is localized then `Movie.ratings` will not work, afaik we will still need to build array of hashes with enum translations

----------

So my main question is; is this feature really needed when we have an option to pass array of hashes to the component similar to what we are already doing with regular rails select_tag?

ps
`defined_enums` are not only place where `form.combobox` will fail in the case of FormObject, but I'll leave that for another discussion or PRs

✌️

**Comment by josefarias** [comment]
Hey @dixpac,

I don't think supporting enums is the problem here. We should just be able to support non-AR objects, as well as AR objects and their enums. PR here: https://github.com/josefarias/hotwire_combobox/pull/151

> if the app is localized then Movie.ratings will not work, afaik we will still need to build array of hashes with enum translations

It works, just not as the implementer would probably want. Per the docs:

> If you want more customization (maybe you want to titlecase the options), you can leverage the library's ability to accept a hash, an array, or an array of arrays as the second argument.

In this case we wouldn't be titlecasing, but rather translating. And that's a use case that would warrant passing a hash or some other data structure.

Our native enum handling is meant as a shortcut for common use cases. Whether it works or not will depend on the implementer's enum setup. Thank means it won't work in every single case, but that's okay! No reason to remove the feature.

Thanks for the question!

**Reply by josefarias** [reply]
Ah unless you meant the combobox works without accessing the defined enums. I'll look into it, can't remember why we added that.
